### PR TITLE
[Feature] Add focal loss

### PR DESCRIPTION
**Pull request by RockeyCoss**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add focal loss

## Modification

Add focal loss

## Experiment Results

We can see through the experiments that:

1. When applying UNet on the medical dataset(including chase, stare, hrf and drive), using focal loss with gamma between 0 to 1 yields relatively good results. 
2. Even with the best choice of gamma, focal loss still can't beat cross entropy loss when comparing their best dice scores during training. When comparing their last dice scores, focal loss outperforms cross entropy loss on stare dataset.
3. Dice loss performs quite good that it outperforms cross entropy loss and focal loss on three datasets.
4. With proper weight, training with multiple losses can yield clear and consistent improvement. For example, focal loss + cross entropy with weight 1:3 or 3:1 and focal loss + dice loss with weight 1:4 yield good results on four datasets.

|exp_num|method|vessel dice best|best index|vessel dice last|last index|
|:--:|:--:|:--:|:--:|:--:|:--:|
|1|pspnet_unet_s5-d16_128x128_40k_chase_db1|0.804|10|0.804|10|
|2|pspnet_unet_s5-d16_128x128_40k_chase_db1_dice|0.797|10|0.797|10|
|3|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0|0.8009|10|0.8009|10|
|4|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5|0.8001|10|0.8001|10|
|5|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce1to1|0.8039|10|0.8039|10|
|6|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce1to2|0.8054|10|0.8054|10|
|7|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce1to3|0.8061|10|0.8061|10|
|8|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce1to4|0.8037|11|0.8037|11|
|9|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce2to1|0.8048|10|0.8048|10|
|10|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce3to1|0.8052|11|0.8052|11|
|11|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_ce4to1|0.8048|9|0.8037|10|
|12|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice1to1|0.8026|10|0.8026|10|
|13|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice1to2|0.8005|10|0.8005|10|
|14|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice1to3|0.7979|11|0.7979|11|
|15|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice1to4|0.7986|12|0.7986|12|
|16|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice2to1|0.8012|10|0.801|11|
|17|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice3to1|0.8052|10|0.8031|11|
|18|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma0.5_dice4to1|0.8076|10|0.8076|10|
|19|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma1|0.7976|10|0.7976|10|
|20|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma1.5|0.793|10|0.793|10|
|21|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma2|0.7877|10|0.7877|10|
|22|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma4|0.7604|10|0.7604|10|
|23|pspnet_unet_s5-d16_128x128_40k_chase_db1_gamma6|0.6452|10|0.6452|10|
|||||||
|24|pspnet_unet_s5-d16_128x128_40k_stare|0.8186|7|0.8097|10|
|25|pspnet_unet_s5-d16_128x128_40k_stare_dice|0.8258|10|0.8258|10|
|26|pspnet_unet_s5-d16_128x128_40k_stare_gamma0|0.8149|7|0.8123|10|
|27|pspnet_unet_s5-d16_128x128_40k_stare_gamma0.5|0.8138|7|0.8129|10|
|28|pspnet_unet_s5-d16_128x128_40k_stare_gamma1|0.8151|10|0.8151|10|
|29|pspnet_unet_s5-d16_128x128_40k_stare_gamma1.5|0.8129|9|0.8095|10|
|30|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce1to1|0.8199|4|0.8146|10|
|31|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce1to2|0.8195|4|0.8149|10|
|32|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce1to3|0.8233|4|0.816|10|
|33|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce1to4|0.8205|7|0.8149|10|
|34|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce2to1|0.8159|8|0.8115|10|
|35|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce3to1|0.8218|7|0.8134|11|
|36|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_ce4to1|0.8221|5|0.8141|10|
|37|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice1to1|0.8263|10|0.8263|10|
|38|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice1to2|0.8274|10|0.8274|10|
|39|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice1to3|0.8298|10|0.8239|11|
|40|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice1to4|0.8262|11|0.8185|12|
|41|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice2to1|0.825|10|0.825|10|
|42|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice3to1|0.8212|10|0.8204|11|
|43|pspnet_unet_s5-d16_128x128_40k_stare_gamma1_dice4to1|0.8248|8|0.8221|11|
|44|pspnet_unet_s5-d16_128x128_40k_stare_gamma2|0.8123|9|0.8092|10|
|45|pspnet_unet_s5-d16_128x128_40k_stare_gamma4|0.7885|10|0.7885|10|
|46|pspnet_unet_s5-d16_128x128_40k_stare_gamma6|0.6574|10|0.6574|10|
|||||||
|47|pspnet_unet_s5-d16_256x256_40k_hrf|0.7992|8|0.7946|12|
|48|pspnet_unet_s5-d16_256x256_40k_hrf_dice|0.8082|10|0.8082|10|
|49|pspnet_unet_s5-d16_256x256_40k_hrf_gamma0|0.7947|8|0.7875|10|
|50|pspnet_unet_s5-d16_256x256_40k_hrf_gamma0.5|0.7959|8|0.79|10|
|51|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1|0.7962|8|0.7889|10|
|52|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1.5|0.7942|8|0.789|10|
|53|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce1to1|0.8059|8|0.7964|10|
|54|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce1to2|0.804|8|0.7938|10|
|55|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce1to3|0.8078|8|0.7962|10|
|56|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce1to4|0.8023|10|0.7936|11|
|57|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce2to1|0.8022|8|0.7897|10|
|58|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce3to1|0.8061|8|0.7917|10|
|59|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_ce4to1|0.8006|8|0.7937|10|
|60|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice1to1|0.8083|8|0.8076|10|
|61|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice1to2|0.8091|10|0.8091|10|
|62|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice1to3|0.8063|8|0.8061|11|
|63|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice1to4|0.8055|12|0.8055|12|
|64|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice2to1|0.8082|11|0.8063|12|
|65|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice3to1|0.8105|8|0.8103|10|
|66|pspnet_unet_s5-d16_256x256_40k_hrf_gamma1_dice4to1|0.8097|8|0.8082|10|
|67|pspnet_unet_s5-d16_256x256_40k_hrf_gamma2|0.7881|9|0.7837|10|
|68|pspnet_unet_s5-d16_256x256_40k_hrf_gamma4|0.7745|9|0.7628|10|
|69|pspnet_unet_s5-d16_256x256_40k_hrf_gamma6|0.6749|9|0.6145|10|
|||||||
|70|pspnet_unet_s5-d16_64x64_40k_drive|0.7926|4|0.7829|10|
|71|pspnet_unet_s5-d16_64x64_40k_drive_dice|0.7951|9|0.795|10|
|72|pspnet_unet_s5-d16_64x64_40k_drive_gamma0|0.7912|4|0.7804|10|
|73|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5|0.7844|4|0.7751|10|
|74|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_cd1to1|0.7865|4|0.7833|10|
|75|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_cd1to2|0.7906|4|0.7807|10|
|76|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_ce1to3|0.792|4|0.7823|10|
|77|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_ce1to4|0.7926|4|0.7839|10|
|78|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_ce2to1|0.796|7|0.785|10|
|79|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_ce3to1|0.7891|6|0.7834|10|
|80|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_ce4to1|0.7942|4|0.7845|10|
|81|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice1to1|0.7935|7|0.7924|10|
|82|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice1to2|0.7968|10|0.7968|10|
|83|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice1to3|0.796|11|0.796|11|
|84|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice1to4|0.7964|10|0.7964|10|
|85|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice2to1|0.7946|4|0.794|10|
|86|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice3to1|0.7968|4|0.7899|12|
|87|pspnet_unet_s5-d16_64x64_40k_drive_gamma0.5_dice4to1|0.8014|9|0.7894|11|
|88|pspnet_unet_s5-d16_64x64_40k_drive_gamma1|0.7811|5|0.7799|10|
|89|pspnet_unet_s5-d16_64x64_40k_drive_gamma1.5|0.7794|5|0.7747|10|
|90|pspnet_unet_s5-d16_64x64_40k_drive_gamma2|0.775|10|0.775|10|
|91|pspnet_unet_s5-d16_64x64_40k_drive_gamma4|0.7579|8|0.7555|10|
|92|pspnet_unet_s5-d16_64x64_40k_drive_gamma6|0.7156|8|0.6892|10|

### CODE REVIEW

**Comment by Junjun2016** [comment]
Hi @RockeyCoss 
Please fix the lint error.

**Comment by Junjun2016** [comment]
Besides, some lines are not covered by unittests.

**Comment by RockeyCoss** [comment]
> 
> 
> Besides, some lines are not covered by unittests.
Can't fix that because the uncovered lines can only be executed when cuda is available.

File: mmseg/models/losses/focal_loss.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        """
+        assert reduction_override in (None, 'none', 'mean', 'sum'), \
+            "AssertionError: reduction should be 'none', 'mean' or " \
+            "'sum'"
+        assert pred.shape == target.shape or \
+               (pred.size(0) == target.size(0) and
+                pred.shape[2:] == target.shape[1:]), \
+               "The shape of pred doesn't match the shape of target"
+
+        none_reduce_shape = pred.shape
```
Comment (RockeyCoss) [inline mmseg/models/losses/focal_loss.py:192]: revise name

File: mmseg/models/losses/focal_loss.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+                    (1 - target)) * pt.pow(gamma)
+    loss = F.binary_cross_entropy_with_logits(
+        pred, target, reduction='none') * focal_weight
+    if weight is not None:
+        if weight.shape != loss.shape:
+            if weight.size(0) == loss.size(0):
+                # For most cases, weight is of shape (num_priors, ),
+                #  which means it does not have the second axis num_class
+                weight = weight.view(-1, 1)
+            else:
```
Comment (RockeyCoss) [inline mmseg/models/losses/focal_loss.py:50]: delete this

File: mmseg/models/losses/focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.ops import sigmoid_focal_loss as _sigmoid_focal_loss
+
+from ..builder import LOSSES
+from .utils import weight_reduce_loss
+
+
+# This method is only for debugging
+def py_sigmoid_focal_loss(pred,
```
Comment (RockeyCoss) [inline mmseg/models/losses/focal_loss.py:13]: ignore index

**Comment by RockeyCoss** [comment]
speed comparision
图片
图片

File: mmseg/models/losses/focal_loss.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+                 class_weight=None,
+                 loss_weight=1.0,
+                 loss_name='loss_focal'):
+        """`Focal Loss <https://arxiv.org/abs/1708.02002>`_
+        Args:
+            use_sigmoid (bool, optional): Whether to the prediction is
+                used for sigmoid or softmax. Defaults to True.
+            gamma (float, optional): The gamma for calculating the modulating
+                factor. Defaults to 2.0.
+            alpha (float, list[float], optional): A balanced form for Focal
```
Comment (Junjun2016) [inline mmseg/models/losses/focal_loss.py:145]: Should add more comments for float type as list type (no different semantic classes or multi-class, only binary class).

File: mmseg/models/losses/focal_loss.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        valid_mask (torch.Tensor ,None): A mask uses 1 to mark the valid
+            samples and uses 0 to mark the ignored samples. Default: None.
+        reduction (str, optional): The method used to reduce the loss into
+            a scalar. Defaults to 'mean'. Options are "none", "mean" and "sum".
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Defaults to None.
+    """
+    # Function.apply does not accept keyword arguments, so the decorator
+    # "weighted_loss" is not applicable
+    float_alpha = alpha if isinstance(alpha, float) else 0.5
```
Comment (Junjun2016) [inline mmseg/models/losses/focal_loss.py:103]: Should add a comment for why we set it to 0.5.

File: mmseg/models/losses/focal_loss.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    if isinstance(alpha, torch.Tensor):
+        final_weight = final_weight * \
+                       (alpha * one_hot_target +
+                        (1 - alpha) * (1 - one_hot_target))
+    if class_weight is not None:
+        final_weight = final_weight * class_weight
+    if valid_mask is not None:
+        final_weight = final_weight * valid_mask
+
+    loss = loss if isinstance(alpha, float) else loss * 2
```
Comment (Junjun2016) [inline mmseg/models/losses/focal_loss.py:123]: Same here for `*2`

**Comment by MengzhangLI** [comment]
@RockeyCoss It seems like some lines are not covered by unit test. Do we need to add unit test?

File: mmseg/models/losses/focal_loss.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                avg_factor=avg_factor)
+
+            if reduction == 'none':
+                # [N, C] -> [C, N]
+                loss_cls = loss_cls.transpose(0, 1)
+                # [C, N] -> [C, B, d1, d2, ...]
+                # original_shape: [B, C, d1, d2, ...]
+                loss_cls = loss_cls.reshape(*original_shape[1:2],
+                                            *original_shape[0:1],
+                                            *original_shape[2:])
```
Comment (Junjun2016) [inline mmseg/models/losses/focal_loss.py:308]:
```suggestion
                loss_cls = loss_cls.reshape(original_shape[1],
                                            original_shape[0],
                                            *original_shape[2:])
```

Comment (RockeyCoss) [reply mmseg/models/losses/focal_loss.py:308]: > @RockeyCoss It seems like some lines are not covered by unit test. Do we need to add unit test?

Those codes need cuda  to run, so it's not possible to test them. However, I have tested them with cuda available enviroment and the `test_same()` in unitests tests if the code will produce the same results when running in cuda available enviroment and cuda unavailable environment.

**Comment by RockeyCoss** [comment]
> 



> @RockeyCoss It seems like some lines are not covered by unit test. Do we need to add unit test?

Those codes need cuda to run, so it's not possible to test them. However, I have tested them with cuda available enviroment and the test_same() in unitests tests if the code will produce the same results when running in cuda available enviroment and cuda unavailable environment.

**Review by Junjun2016 — APPROVED** [review]

File: tests/test_models/test_losses/test_focal_loss.py [hunk 8]
```diff
@@ -0,0 +1,1618 @@
...
+                       [[1.986406743526e-02, 2.876800950617e-03],
+                        [3.475922346115e-02, 7.408759556711e-03]]],
+                      [[[6.621975335293e-04, 5.068928934634e-03],
+                        [3.801103215665e-03, 1.346173667116e-04]],
+                       [[1.120079588145e-02, 1.886526471935e-03],
+                        [2.504604868591e-02, 2.119205333292e-02]],
+                       [[3.766115009785e-02, 6.910999864340e-02],
+                        [3.804825479165e-03, 1.751711219549e-01]]]]),
+        torch.tensor([[[[3.154235659167e-03, 4.470038227737e-03],
+                        [7.556425873190e-03, 4.038503859192e-03]],
```
Comment (xvjiarui) [inline tests/test_models/test_losses/test_focal_loss.py:254]: We may use smaller array to test.

File: tests/test_models/test_losses/test_focal_loss.py [hunk 9]
```diff
@@ -0,0 +1,1618 @@
...
+                            index += 1
+                            focal_loss.class_weight = class_weight
+                            focal_loss.alpha = alpha
+                            cpu_result = focal_loss(
+                                fake_pred,
+                                target,
+                                weight=weight,
+                                ignore_index=ignore_index,
+                                reduction_override=reduce)
+                            assert torch.allclose(cpu_result, gpu_result)
```
Comment (xvjiarui) [inline tests/test_models/test_losses/test_focal_loss.py:1618]: The nest is too deep.
A single test for each case (covering all the branches) is sufficient.

Comment (RockeyCoss) [reply tests/test_models/test_losses/test_focal_loss.py]: ok

File: mmseg/models/losses/focal_loss.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Defaults to None.
+    """
+    # Function.apply does not accept keyword arguments, so the decorator
+    # "weighted_loss" is not applicable
+
+    # _sigmoid_focal_loss doesn't accept alpha of list type. Therefore, if
+    # a list is given, we set float_alpha as 0.5. This means setting equal
+    # weight for foreground class and background class.
+    float_alpha = alpha if isinstance(alpha, float) else 0.5
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:107]:
```suggestion
   assert isinstance(alpha, float)
```
This way may make user input float explicitly, avoid ambiguity

Comment (RockeyCoss) [reply mmseg/models/losses/focal_loss.py:107]: This won't implicitly change the input alpha of type list to float 0.5. In order to use the cuda version focal loss. I have to set alpha as 0.5. Then in the post-processing process, I offset the effect of setting alpha as 0.5 and use the user-provided alpha(a list) to regulate the loss.(line 118-121) Therefore, providing an alpha of list type is acceptable.

Comment (RockeyCoss) [reply mmseg/models/losses/focal_loss.py:107]: What's more, this function is only called by class FocalLoss, who makes sure that only alpha of type float and tensor will be input to the function.

File: mmseg/models/losses/focal_loss.py [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+                        -1, 1)
+                calculate_loss_func = py_sigmoid_focal_loss
+
+            loss_cls = self.loss_weight * calculate_loss_func(
+                pred,
+                target,
+                one_hot_target,
+                weight,
+                gamma=self.gamma,
+                alpha=self.alpha if isinstance(
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:293]: Could we pass original `alpha` into the function? And deal with the list and float inside?

File: mmseg/models/losses/focal_loss.py [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+                               float_alpha, None, 'none')
+    final_weight = torch.ones(1, pred.size(1)).type_as(loss)
+    if weight is not None:
+        if (weight.shape != loss.shape and weight.size(0) == loss.size(0)):
+            # For most cases, weight is of shape (N, ),
+            # which means it does not have the second axis num_class
+            weight = weight.view(-1, 1)
+        assert weight.dim() == loss.dim()
+        final_weight = final_weight * weight
+    if isinstance(alpha, torch.Tensor):
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:118]: Could we pass original alpha into the function? And deal with the list and float inside?

File: mmseg/models/losses/focal_loss.py [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, C), C is the number
+            of classes.
+        target (torch.Tensor): The learning label of the prediction. It's shape
+            should be (N, )
+        one_hot_target (torch.Tensor): The learning label with shape (N, C)
+        weight (torch.Tensor, optional): Sample-wise loss weight.
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 2.0.
+        alpha (float, torch.Tensor, optional): A balanced form for Focal Loss.
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:90]: Could we pass original alpha into the function? And deal with the list and float inside?

File: mmseg/models/losses/focal_loss.py (hunk 10 above)
Comment (xvjiarui) [reply mmseg/models/losses/focal_loss.py:107]: I see. We may explain it in the comments. And maybe rename `float_alpha` to `pseudo_alpha`?

Comment (RockeyCoss) [reply mmseg/models/losses/focal_loss.py:107]: ok

File: mmseg/models/losses/focal_loss.py [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+    """
+    # Function.apply does not accept keyword arguments, so the decorator
+    # "weighted_loss" is not applicable
+    final_weight = torch.ones(1, pred.size(1)).type_as(pred)
+    if isinstance(alpha, list):
+        # _sigmoid_focal_loss doesn't accept alpha of list type. Therefore, if
+        # a list is given, we set the input alpha as 0.5. This means setting
+        # equal weight for foreground class and background class. By
+        # multiplying the loss by 2, the effect of setting alpha as 0.5 is
+        # offset. The alpha of type list is used to regulate the loss in the
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:111]:
```suggestion
        # undone. The alpha of type list is used to regulate the loss in the
```

File: mmseg/models/losses/focal_loss.py [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+
+            loss_cls = self.loss_weight * calculate_loss_func(
+                pred,
+                target,
+                one_hot_target,
+                weight,
+                gamma=self.gamma,
+                alpha=self.alpha,
+                class_weight=None if not self.class_weight else
+                pred.new_tensor(self.class_weight),
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:296]: We may also convert inside the function

File: mmseg/models/losses/focal_loss.py [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+                          avg_factor=None):
+    """PyTorch version of `Focal Loss <https://arxiv.org/abs/1708.02002>`_.
+
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, C), C is the
+            number of classes
+        target (torch.Tensor): The learning label of the prediction with
+            shape (N, C)
+        one_hot_target (None): Placeholder. It should be None.
+        weight (torch.Tensor, optional): Sample-wise loss weight.
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:31]: We may also convert inside the function

File: mmseg/models/losses/focal_loss.py [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 2.0.
+        alpha (float, list, optional): A balanced form for Focal Loss.
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:34]:
```suggestion
        alpha (float | list[float], optional): A balanced form for Focal Loss.
```

File: mmseg/models/losses/focal_loss.py [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+            should be (N, )
+        one_hot_target (torch.Tensor): The learning label with shape (N, C)
+        weight (torch.Tensor, optional): Sample-wise loss weight.
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 2.0.
+        alpha (float, list, optional): A balanced form for Focal Loss.
+            Defaults to 0.5.
+        class_weight (torch.Tensor, optional): Weight of each class.
+            Defaults to None.
+        valid_mask (torch.Tensor ,None): A mask uses 1 to mark the valid
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:96]:
```suggestion
        valid_mask (torch.Tensor, optional): A mask uses 1 to mark the valid
```

File: mmseg/models/losses/focal_loss.py [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+            shape (N, C)
+        one_hot_target (None): Placeholder. It should be None.
+        weight (torch.Tensor, optional): Sample-wise loss weight.
+        gamma (float, optional): The gamma for calculating the modulating
+            factor. Defaults to 2.0.
+        alpha (float, list, optional): A balanced form for Focal Loss.
+            Defaults to 0.5.
+        class_weight (torch.Tensor, optional): Weight of each class.
+            Defaults to None.
+        valid_mask (torch.Tensor ,None): A mask uses 1 to mark the valid
```
Comment (xvjiarui) [inline mmseg/models/losses/focal_loss.py:38]:
```suggestion
        valid_mask (torch.Tensor, optional): A mask uses 1 to mark the valid
```

**Review by xvjiarui — APPROVED** [review]

**Comment by Zihang-Wei** [comment]
Hello, how should I use the focal loss inside the config file? It pops up this strange message for me...
image
In the config file: 
```
model = dict(
    test_cfg=dict(crop_size=(128, 128), stride=(85, 85)), 
    auxiliary_head=dict(
        num_classes=3,
        loss_decode=[
            # dict(type='CrossEntropyLoss', loss_name='loss_ce'),
            # dict(type='DiceLoss', loss_name='loss_dice', loss_weight=3.0),
            dict(type='FocalLoss', loss_name='loss_focal', loss_weight=3.0, gamma=2.0, alpha=[0.2, 0.1, 0.7])
            ]),
    decode_head=dict(
        num_classes=3, 
        loss_decode=[
            # dict(type='CrossEntropyLoss', loss_name='loss_ce', loss_weight=1.0),
            # dict(type='DiceLoss', loss_name='loss_dice', loss_weight=3.0),
            dict(type='FocalLoss', loss_name='loss_focal', loss_weight=3.0, gamma=2.0, alpha=[0.2, 0.1, 0.7]),
            # dict(type='LovaszLoss', loss_name='loss_lovasz', loss_weight=3.0)
            ]))
```
@RockeyCoss @Junjun2016 @MengzhangLI @xvjiarui @kahkeng

**Comment by RockeyCoss** [comment]
> Hello, how should I use the focal loss inside the config file? It pops up this strange message for me... image In the config file:
> 
> ```
> model = dict(
>     test_cfg=dict(crop_size=(128, 128), stride=(85, 85)), 
>     auxiliary_head=dict(
>         num_classes=3,
>         loss_decode=[
>             # dict(type='CrossEntropyLoss', loss_name='loss_ce'),
>             # dict(type='DiceLoss', loss_name='loss_dice', loss_weight=3.0),
>             dict(type='FocalLoss', loss_name='loss_focal', loss_weight=3.0, gamma=2.0, alpha=[0.2, 0.1, 0.7])
>             ]),
>     decode_head=dict(
>         num_classes=3, 
>         loss_decode=[
>             # dict(type='CrossEntropyLoss', loss_name='loss_ce', loss_weight=1.0),
>             # dict(type='DiceLoss', loss_name='loss_dice', loss_weight=3.0),
>             dict(type='FocalLoss', loss_name='loss_focal', loss_weight=3.0, gamma=2.0, alpha=[0.2, 0.1, 0.7]),
>             # dict(type='LovaszLoss', loss_name='loss_lovasz', loss_weight=3.0)
>             ]))
> ```
> 
> @RockeyCoss @Junjun2016 @MengzhangLI @xvjiarui @kahkeng

Hello @Zihang-Wei ! Did you install mmcv-full? By the way, we would appreciate it if you could report this problem in the [issue](https://github.com/open-mmlab/mmsegmentation/issues). Because by doing this, the problem can be viewed by more people.

**Comment by Zihang-Wei** [comment]
Yes, inside my dockerfile, there is:
image

**Comment by Zihang-Wei** [comment]
However since inside VScode, the mmcv package is inside /opt/venv/packages/... therefore the callbacks won't step into the function imported from mmcv.ops

**Pull request merged**